misc/testerina/modules/testerina-core/src/main/ballerina/concurrentExecuter.bal

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2024 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+isolated function executeTestIso(TestFunction testFunction, DataProviderReturnType? testFunctionArgs) {
+    if !isTestReadyToExecute(testFunction, testFunctionArgs) {
+        return;
+    }
+
+    executeBeforeGroupFunctionsIso(testFunction);
+    executeBeforeEachFunctionsIso();
+
+    boolean shouldSkipDependents = false;
+    if !isSkipFunction(testFunction) {
+        if (isDataDrivenTest(testFunctionArgs)) {
+            executeDataDrivenTestSetIso(testFunction, testFunctionArgs);
+        } else {
+            shouldSkipDependents = executeNonDataDrivenTestIso(testFunction, testFunctionArgs);
+        }
+    } else {
+        reportData.onSkipped(name = testFunction.name, testType = getTestType(testFunctionArgs));
+        shouldSkipDependents = true;
+    }
+    testFunction.groups.forEach('group => groupStatusRegistry.incrementExecutedTest('group));
+    executeAfterEachFunctionsIso();
+    executeAfterGroupFunctionsIso(testFunction);
+    finishTestExecution(testFunction, shouldSkipDependents);
+}
+
+isolated function executeBeforeGroupFunctionsIso(TestFunction testFunction) {
+    foreach string 'group in testFunction.groups {
+        TestFunction[]? beforeGroupFunctions = beforeGroupsRegistry.getFunctions('group);
+        if beforeGroupFunctions != () && !groupStatusRegistry.firstExecuted('group) {
+            handleBeforeGroupOutput(testFunction, 'group, executeFunctionsIso(beforeGroupFunctions, getShouldSkip()));
+        }
+    }
+}
+
+isolated function executeBeforeEachFunctionsIso() =>
+    handleBeforeEachOutput(executeFunctionsIso(beforeEachRegistry.getFunctions(), getShouldSkip()));
+
+isolated function executeDataDrivenTestSetIso(TestFunction testFunction, DataProviderReturnType? testFunctionArgs) {
+    string[] keys = [];
+    AnyOrErrorOrReadOnlyType[][] values = [];
+    TestType testType = prepareDataSet(testFunctionArgs, keys, values);
+    map<future> futures = {};
+
+    while keys.length() != 0 {
+        string key = keys.remove(0);
+        ReadOnlyType[] value = <ReadOnlyType[]>values.remove(0);
+        future<()> futureResult = start prepareDataDrivenTestIso(testFunction, key, value.clone(), testType);
```

Comment:
```suggestion
        future<()> futureResult = start prepareDataDrivenTestIso(testFunction, key, [...value], testType);
```
@MaryamZi rather than cloning isn't it better to create a new array?

Reply:
I guess one slight advantage with spread is that we avoid unnecessary is checks or a clone invocations (although they will return the values without cloning).